app/src/main/java/de/westnordost/streetcomplete/controls/MainMenuButtonFragment.kt

```diff
@@ -0,0 +1,104 @@
+
+    /* --------------------------------------- Lifecycle ---------------------------------------- */
+
+    init {
+        Injector.applicationComponent.inject(this)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        view.mainMenuButton.setOnClickListener { onClickMainMenu() }
+    }
+
+    /* ------------------------------------------------------------------------------------------ */
+
+    internal fun onClickMainMenu() {
+        context?.let { MainMenuDialog(it, this::onClickDownload).show() }
+    }
+
+    /* ------------------------------------ Download Button  ------------------------------------ */
+
+    private fun onClickDownload() {
+        if (isConnected()) downloadDisplayedArea()
+        else context?.toast(R.string.offline)
+    }
+
+    private fun isConnected(): Boolean {
+        val connectivityManager = context?.getSystemService<ConnectivityManager>()
+        val activeNetworkInfo = connectivityManager?.activeNetworkInfo
+        return activeNetworkInfo != null && activeNetworkInfo.isConnected
+    }
+
+    private fun downloadDisplayedArea() {
+        val displayArea = (requireParentFragment() as MainFragment).mapFragment?.getDisplayedArea()
+        if (displayArea == null) {
+            context?.toast(R.string.cannot_find_bbox_or_reduce_tilt, Toast.LENGTH_LONG)
+        } else {
+            val enclosingBBox = displayArea.asBoundingBoxOfEnclosingTiles(ApplicationConstants.QUEST_TILE_ZOOM)
+            val areaInSqKm = enclosingBBox.area() / 1000000
+            if (areaInSqKm > ApplicationConstants.MAX_DOWNLOADABLE_AREA_IN_SQKM) {
+                context?.toast(R.string.download_area_too_big, Toast.LENGTH_LONG)
+            } else {
+                if (downloadController.isPriorityDownloadInProgress) {
+                    context?.let {
+                        AlertDialog.Builder(it)
+                            .setMessage(R.string.confirmation_cancel_prev_download_title)
+                            .setPositiveButton(R.string.confirmation_cancel_prev_download_confirmed) { _, _ ->
+                                downloadAreaConfirmed(enclosingBBox)
+                            }
+                            .setNegativeButton(R.string.confirmation_cancel_prev_download_cancel, null)
+                            .show()
+                    }
+                } else {
+                    downloadAreaConfirmed(enclosingBBox)
+                }
+            }
+        }
+    }
+
+    private fun downloadAreaConfirmed(bbox: BoundingBox) {
+        var bbox = bbox
+        val areaInSqKm = bbox.area() / 1000000
+        // below a certain threshold, it does not make sense to download, so let's enlarge it
+        if (areaInSqKm < ApplicationConstants.MIN_DOWNLOADABLE_AREA_IN_SQKM) {
+            val cameraPosition = (requireParentFragment() as MainFragment).mapFragment?.cameraPosition
```

Comment:
same here

Reply:
Maybe that part and the deciding if it is too large or too small should go into whoever implements the `getDownloadArea`, what do you think?